proposals/nnnn-span-access-shared-contiguous-storage.md

```diff
@@ -0,0 +1,821 @@
+
+## Proposed solution
+
+#### `Span`
+
+`Span` will allow sharing the contiguous internal representation of a type, by providing access to a borrowed view of an interval of contiguous memory. `Span` does not copy the underlying data: it instead relies on a guarantee that the original container cannot be modified or destroyed while the `Span` exists. In the prototype that accompanies this first proposal, `Span`s will be constrained to closures from which they structurally cannot escape. Later, we will introduce a lifetime dependency between a `Span` and the binding of the type vending it, preventing its escape from the scope where it is valid for use. Both of these approaches guarantee temporal safety. `Span` also performs bounds-checking on every access to preserve spatial safety. Additionally `Span` always represents initialized memory, preserving the definite initialization guarantee.
+
+`Span` is intended as the currency type for local processing over values in contiguous memory. It is a replacement for many API currently using `Array`, `UnsafeBufferPointer`, `Foundation.Data`, etc., that do not need to escape the owning container.
+
+A `Span` provided by a container represents a borrow of that container. `Span` can therefore provide simultaneous access to a non-copyable container. It can also help avoid unwanted copies of copyable containers. Note that `Span` is not a replacement for a copyable container with owned storage; see [future directions](#Directions) for more details ([Resizable, contiguously-stored, untyped collection in the standard library](#Bytes))
+
+#### `RawSpan`
+
+`RawSpan` allows sharing contiguous memory representing values which may be heterogeneously-typed, such as memory intended for parsing. It makes the same safety guarantees as `Span`. Since it is a fully concrete type, it can achieve great performance in debug builds of client code as well as straightforward performance in library code.
+
+A `RawSpan` can be obtained from containers of `BitwiseCopyable` elements, as well as be initialized directly from an instance of `Span<T: BitwiseCopyable>`.
+
+## <a name="Design"></a>Detailed design
+
+`Span<Element>` is a simple representation of a region of initialized memory.
+
+```swift
+@frozen
+public struct Span<Element: ~Copyable & ~Escapable>: Copyable, ~Escapable {
+  internal var _start: UnsafePointer<Element>
+  internal var _count: Int
+}
+```
+
+It provides a buffer-like interface to the elements stored in that span of memory:
+
+```swift
+extension Span where Element: ~Copyable & ~Escapable {
+  public var count: Int { get }
+  public var isEmpty: Bool { get }
+
+  public subscript(_ position: Int) -> Element { _read }
```

Comment:
> public subscript(_ position: Int) -> Element { _read }

This is not, in fact, the API that we want long term; it won't be compatible with lifetime-dependent values that are derived from the span.

We really want some new spelling for `~Copyable` projections, like:

> public subscript(_ position: Int) -> Element { borrow }

For now, we could say:

> public subscript(_ position: Int) -> Element { unsafeAddress }

Either way, we end up with an extra mangling for this API that may become vestigial in the future. But with `unsafeAddress`, at least we have the full functionality of some future, safe "borrow accessor".

@jckarter what do you think?

Reply:
The future directions attempts to explain that the `_read` accessor is a placeholder. I expect the implementation of such placeholders to be non-ABI until there is a final spelling. I could add a reference to the future directions item after this.

Reply:
Ok. I realize the plan is to initially implement Span.subscript via `_read`. I'll just point out that it will limit usability. `unsafeAddress` will at least support the same use cases that a `borrow` accessor would support.